File: LeetCode/Python Solutions/Binary Subarrays With Sum/binary.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        prefixSumCount = collections.defaultdict(int)  # Map to store count of prefix sums
        runningSum = 0
        count = 0
        
        for num in nums:
            # Increment the running sum
            runningSum += num
            # If runningSum equals goal, increment count
            if runningSum == goal:
                count += 1
            # Add to count the number of times (runningSum - goal) has occurred
            count += prefixSumCount[runningSum - goal]
            # Increment the count of runningSum in the map
            prefixSumCount[runningSum] += 1
        
        return count

File: LeetCode/Java Solutions/Binary Subarrays With Sum/binary.java
import java.util.HashMap;

class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        HashMap<Integer, Integer> prefixSumCount = new HashMap<>();  // Map to store count of prefix sums
        int runningSum = 0;
        int count = 0;
        
        for (int num : nums) {
            // Increment the running sum
            runningSum += num;
            // If runningSum equals goal, increment count
            if (runningSum == goal) {
                count++;
            }
            // Add to count the number of times (runningSum - goal) has occurred
            count += prefixSumCount.getOrDefault(runningSum - goal, 0);
            // Increment the count of runningSum in the map
            prefixSumCount.put(runningSum, prefixSumCount.getOrDefault(runningSum, 0) + 1);
        }
        
        return count;
    }
}


